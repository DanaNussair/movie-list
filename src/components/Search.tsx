import {
  useState,
  useEffect,
  useRef,
  MutableRefObject,
  useContext,
} from "react";
import {
  SearchInput,
  SearchContainer,
  SearchWrapper,
  InputIcon,
  ResultItems,
  NoResultsItem,
  SkeletonWrapper,
  SkeletonPoster,
  SkeletionInfo,
  SkeletionTitle,
  SkeletionYear,
} from "../styles/SearchStyles";
import { Magnifier } from "../assets";
import { searchMovies } from "../api/MovieAPI";
import { MovieItemType } from "../types";
import MovieItem from "./MovieItem";
import { AppError, ErrorContext } from "../contexts/ErrorContext";

const Search = () => {
  const [searchResults, setSearchResults] = useState<MovieItemType[] | null>(
    null
  );
  const [searchTerm, setSearchTerm] = useState<string>("");
  const [isListHidden, setIsListHidden] = useState<boolean>(!searchTerm);
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const searchRef = useRef() as MutableRefObject<HTMLDivElement>;

  const { setError } = useContext(ErrorContext);

  useEffect(() => {
    // Will call the search API only when the user has stopped typing, to prevent multiple meaningless calls
    const timeOutId = setTimeout(() => {
      searchTerm
        ? searchMovies(searchTerm, setData)
        : setData({
            data: null,
            isHidden: true,
            isLoading: false,
          });
    }, 500);

    return () => clearTimeout(timeOutId);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchTerm]);

  useEffect(() => {
    document.addEventListener("mousedown", handleClickOutside);
    return () => document.removeEventListener("mousedown", handleClickOutside);
  });

  const setData = ({
    data,
    isHidden = true,
    isLoading = false,
    error = null,
  }: {
    data: MovieItemType[] | null;
    isHidden?: boolean;
    isLoading?: boolean;
    error?: AppError | null;
  }) => {
    setSearchResults(data);
    setIsListHidden(isHidden);
    setIsLoading(isLoading);
    setError(error);
  };

  const handleClickOutside = (e: any) => {
    if (!searchRef.current.contains(e.target) || !searchTerm) {
      setIsListHidden(true);
    } else {
      setIsListHidden(false);
    }
  };

  const displaySkeletonLoader = () => {
    return (
      <ResultItems>
        <SkeletonWrapper>
          <SkeletonPoster />
          <SkeletionInfo>
            <SkeletionTitle />
            <SkeletionYear />
          </SkeletionInfo>
        </SkeletonWrapper>
      </ResultItems>
    );
  };

  const displaySearchResults = () => {
    if (isLoading) {
      return displaySkeletonLoader();
    }
    if (!isListHidden) {
      return (
        <ResultItems>
          {searchResults?.length ? (
            searchResults.map((movie) => (
              <MovieItem
                movie={movie}
                searchTerm={searchTerm}
                key={movie.imdbID}
              />
            ))
          ) : (
            <NoResultsItem>No results found</NoResultsItem>
          )}
        </ResultItems>
      );
    } else return null;
  };

  return (
    <SearchContainer>
      <SearchWrapper
        ref={searchRef}
        onClick={(event) => handleClickOutside(event)}
      >
        <InputIcon>{Magnifier}</InputIcon>
        <SearchInput
          type="text"
          placeholder="Search"
          name="Search"
          onChange={(event) => setSearchTerm(event.target.value)}
          searchResults={searchResults}
          isListHidden={isListHidden}
        />
        {displaySearchResults()}
      </SearchWrapper>
    </SearchContainer>
  );
};

export default Search;
