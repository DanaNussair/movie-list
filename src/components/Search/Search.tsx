import {
  useState,
  useEffect,
  useRef,
  MutableRefObject,
  useContext,
} from "react";
import {
  SearchInput,
  SearchContainer,
  SearchWrapper,
  InputIcon,
  ResultItems,
  NoResultsItem,
  SkeletonWrapper,
  SkeletonPoster,
  SkeletionInfo,
  SkeletionTitle,
  SkeletionYear,
} from "./SearchStyles";
import { Magnifier } from "../../assets";
import { searchMovies } from "../../api/MovieAPI";
import { MovieItemType } from "../../types";
import MovieItem from "../MovieItem/MovieItem";
import { AppError, ErrorContext } from "../../contexts/ErrorContext";

const Search = () => {
  const [searchResults, setSearchResults] = useState<MovieItemType[] | null>(
    null
  );
  const [searchTerm, setSearchTerm] = useState<string>("");
  const [isListHidden, setIsListHidden] = useState<boolean>(!searchTerm);
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const searchRef = useRef() as MutableRefObject<HTMLDivElement>;

  const { setError } = useContext(ErrorContext);

  useEffect(() => {
    // Will call the search API only when the user has stopped typing, to prevent multiple meaningless calls
    const timeOutId = setTimeout(() => {
      handleSearch();
    }, 500);

    return () => clearTimeout(timeOutId);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchTerm]);

  useEffect(() => {
    document.addEventListener("mousedown", handleClickOutside);
    return () => document.removeEventListener("mousedown", handleClickOutside);
  });

  const setData = ({
    data = null,
    isHidden = true,
    isLoading = false,
    error = null,
  }: {
    data?: MovieItemType[] | null;
    isHidden?: boolean;
    isLoading?: boolean;
    error?: AppError | null;
  }) => {
    setSearchResults(data);
    setIsListHidden(isHidden);
    setIsLoading(isLoading);
    setError(error);
  };

  const handleSearch = async () => {
    if (searchTerm) {
      setData({ isHidden: false, isLoading: true });
      const response = await searchMovies(searchTerm);
      setData({
        ...response,
        isHidden: false,
        isLoading: false,
      });
    } else {
      setData({
        data: null,
        isHidden: true,
        isLoading: false,
      });
    }
  };

  const handleClickOutside = (e: any) => {
    if (!searchRef.current.contains(e.target) || !searchTerm) {
      setIsListHidden(true);
    } else {
      setIsListHidden(false);
    }
  };

  const displaySkeletonLoader = () => {
    return (
      <ResultItems data-testid="loading-skeleton">
        <SkeletonWrapper>
          <SkeletonPoster />
          <SkeletionInfo>
            <SkeletionTitle />
            <SkeletionYear />
          </SkeletionInfo>
        </SkeletonWrapper>
      </ResultItems>
    );
  };

  const displaySearchResults = () => {
    if (isLoading) {
      return displaySkeletonLoader();
    }
    if (!isListHidden) {
      return (
        <ResultItems data-testid="result-items">
          {searchResults?.length ? (
            searchResults.map((movie) => (
              <MovieItem
                movie={movie}
                searchTerm={searchTerm}
                key={movie.imdbID}
              />
            ))
          ) : (
            <NoResultsItem data-testid="no-results">
              No results found
            </NoResultsItem>
          )}
        </ResultItems>
      );
    } else return null;
  };

  return (
    <SearchContainer>
      <SearchWrapper
        ref={searchRef}
        onClick={(event) => handleClickOutside(event)}
      >
        <InputIcon>{Magnifier}</InputIcon>
        <SearchInput
          type="text"
          placeholder="Search"
          name="Search"
          onChange={(event) => setSearchTerm(event.target.value)}
          searchResults={searchResults}
          isListHidden={isListHidden}
          isLoading={isLoading}
          data-testid="search-input"
        />
        {displaySearchResults()}
      </SearchWrapper>
    </SearchContainer>
  );
};

export default Search;
