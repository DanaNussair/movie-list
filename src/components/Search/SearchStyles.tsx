import styled, { keyframes } from "styled-components";
import { MovieItemType } from "../../types";

interface Props {
  ref?: React.MutableRefObject<HTMLDivElement>;
}
export const SearchContainer = styled.div`
  display: flex;
  background-color: #001a76;
  height: 151px;
  justify-content: center;
  align-items: center;
`;

export const SearchWrapper = styled.div<Props>`
  position: relative;
  width: 35%;
  height: 60px;
`;

interface SearchInputProps {
  searchResults: MovieItemType[] | null;
  isListHidden: boolean;
  isLoading: boolean;
}

export const SearchInput = styled.input<SearchInputProps>`
  position: absolute;
  box-sizing: border-box;
  height: 100%;
  width: 100%;
  padding-left: 60px;
  padding-right: 20px;
  border-radius: 41px;
  border: 1px solid #001a76;
  border-bottom: none;
  font-size: 20px;
  font-weight: bold;

  ::placeholder {
    color: #090000;
    opacity: 0.14;
  }
  :focus,
  :hover {
    outline: none;
  }

  ${({ searchResults, isLoading, isListHidden }) =>
    ((searchResults && !isListHidden) || isLoading) &&
    `border-radius: 30px 30px 0px 0px;`}
`;

export const InputIcon = styled.div`
  position: absolute;
  margin: 16px;
  width: 29px;
  height: 29px;
  z-index: 1;
`;

export const ResultItems = styled.div`
  position: relative;
  box-sizing: border-box;
  top: 60px;
  max-height: 292px;
  overflow-y: overlay;
  border-radius: 0px 0px 41px 41px;
  border: 1px solid #001a76;
  background-color: white;

  ::-webkit-scrollbar {
    border-radius: 10px;
  }
  ::-webkit-scrollbar-button {
    height: 25px;
  }
  ::-webkit-scrollbar-thumb {
    background-color: #879eed;
    border-radius: 10px;
  }
`;

export const NoResultsItem = styled.div`
  display: flex;
  height: 100px;
  justify-content: center;
  align-items: center;
  font-weight: bold;
  font-size: 18px;
  color: #9c9c9c;
`;

export const SkeletonWrapper = styled.div`
  display: flex;
  flex-direction: row;
  gap: 10px;
  padding: 13px;
`;

const AnimateBG = keyframes`
  to {
    background-position: 300px 0, 0 0, 0 190px, 50px 195px;
  }
`;

export const SkeletonPoster = styled.div`
  height: 70px;
  width: 70px;
  border-radius: 10px;
  background-size: 300px 300px;
  background-image: linear-gradient(90deg, #e2e2e2 50%, #ebebeb 100%);
  animation: ${AnimateBG} 1s ease infinite;
`;

export const SkeletionInfo = styled.div`
  width: 75%;
  display: flex;
  flex-direction: column;
  gap: 14px;
`;

export const SkeletionTitle = styled.div`
  height: 24px;
  border-radius: 10px;
  background-color: #e2e2e2;
  background-size: 300px 300px;
  background-image: linear-gradient(90deg, #e2e2e2 0%, #ebebeb 100%);
  animation: ${AnimateBG} 1s ease infinite;
`;

export const SkeletionYear = styled.div`
  width: 40%;
  height: 24px;
  border-radius: 10px;
  background-color: #e2e2e2;
  background-size: 300px 300px;
  background-image: linear-gradient(90deg, #e2e2e2 0%, #ebebeb 100%);
  animation: ${AnimateBG} 1s ease infinite;
`;
