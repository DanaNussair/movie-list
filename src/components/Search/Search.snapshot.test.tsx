import renderer from "react-test-renderer";
import Search from "./Search";

describe("<Search />", () => {
  it("renders search component", () => {
    let tree = renderer.create(<Search />);

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
