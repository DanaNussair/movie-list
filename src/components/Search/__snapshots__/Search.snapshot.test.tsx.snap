// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> renders search component 1`] = `
<div
  className="sc-bcXHqe hPXerF"
>
  <div
    className="sc-gswNZR gEhWMn"
    onClick={[Function]}
  >
    <div
      className="sc-hLBbgP dWzqHG"
    >
      <img
        alt="magnifier"
        src="magnifier.svg"
      />
    </div>
    <input
      className="sc-dkrFOg bNsDVa"
      data-testid="search-input"
      name="Search"
      onChange={[Function]}
      placeholder="Search"
      type="text"
    />
  </div>
</div>
`;
